byway/src/main/webapp/routepage.js

```diff
@@ -12,117 +12,198 @@
   document.addEventListener('DOMContentLoaded', loadData);
 } else {  // `DOMContentLoaded` has already fired
   loadData();
 }
-
-/** Used to restore stops and recommendations upon load or refresh */
+ 
+/** Used to restore route and recommendations upon load or refresh */
 function loadData(){
   getRecsOnload();
-  getStopsOnload();
+  getRouteOnload();
 }
-
+ 
 /** Initializes map on the page */
 function initMap() {
-  const directionsService = new google.maps.DirectionsService();
-  const directionsRenderer = new google.maps.DirectionsRenderer();
-  let start = new google.maps.LatLng(37.7699298, -122.4469157);
-  let end = new google.maps.LatLng(37.7683909618184, -122.51089453697205);
+  directionsService = new google.maps.DirectionsService();
+  directionsRenderer = new google.maps.DirectionsRenderer();
+ 
   let mapOptions = {
     zoom: 14,
-    center: start
+    center: new google.maps.LatLng(40.730610, -73.935242) // coordinates of NYC
   }
   const map = new google.maps.Map(document.getElementById('map'), mapOptions);
   directionsRenderer.setMap(map);
-  document.getElementById("route").addEventListener("click", function() {
-    calcRoute(directionsService, directionsRenderer, start, end);
-  });
+  
 }
 
-/** 
- * Displays route containing waypoints overtop the map.
- * TODO: Add in code from routeDir to change route in real-time
- * based on what's in the schedule panel
- * @param {DirectionsService} directionsService object that communicates with the GMaps API service
- * @param {DirectionsRenderer} directionsRenderer object that renders display results on the map
- * @param {String} start starting point of route
- * @param {String} end ending point of route. TODO: In the final implementation, just have 
- * start as start and end are the same
- */
-function calcRoute(directionsService, directionsRenderer, start, end) {
+/** Displays route containing waypoints overtop the map. */
+function calcRoute() {
   let request = {
     origin:  start,
     destination: end,
-    travelMode: 'DRIVING'
+    travelMode: 'DRIVING',
+    waypoints:  route.map(waypoint => ({location: waypoint})),
+    optimizeWaypoints: true
   };
   directionsService.route(request, function(response, status) {
     if (status == 'OK') {
       directionsRenderer.setDirections(response);
+      orderWaypoints(response);
+      updateDistanceTime(response);
     } else {
       window.alert("Could not calculate route due to: " + status);
     }
+    updateRoute();
   });
 }
 
-/** Clear the stops panel in the html */
-function clearStops(){
-  const stopList = document.getElementById('stop-list');
-  if(stopList != null){
-    stopList.innerText = "";
+/** Add the starting location back to the schedule panel
+ *  TODO: Delete markers for recommended stops not selected.
+ *  TODO: Disable usage after? Don't want to keep adding to list. 
+ */
+function generateRoute() {
+  route.splice(0,0,start);
+  route.splice(route.length, 0, end);
```

Comment:
What is this doing? Try to keep the state of the `route` list consistent, that is, make it so it either always contains the start and end, or never contains the start and end. Having a function which adds the start and end without clear tracking of whether the start and end are currently included in the list is error prone, because it becomes very easy to run this add operation extra times even when the list already contains the start/end.

Reply:
I just wanted to add back the start/end location to the schedule panel when the user has finalized the stops they want to visit,  and the way I did this was add them back to the route. It's more just a visual thing. Yeah, I didn't add anything to prevent this action from being taken multiple times. Do you have any suggestions as to how to improve this?

Reply:
Just add the start/end when creating elements on the page, without changing the underlying data stored in the page-wide variables.

Reply:
Oh I see, I changed generateRoute() to reflect this.